products/llbuildSwift/BuildDBBindings.swift

```diff
@@ -0,0 +1,345 @@
+}
+
+extension RuleResult: CustomReflectable {
+    public var customMirror: Mirror {
+        // We exclude the result pointer from the reflection
+        let properties: DictionaryLiteral<String, Any> = [
+            "signature": signature,
+            "computedAt": computedAt,
+            "builtAt": builtAt,
+            "dependencies": dependencies,
+            "value": value
+        ]
+        
+        return Mirror(RuleResult.self, children: properties, displayStyle: .class, ancestorRepresentation: .generated)
+    }
+}
+
+/// Private class for easier handling of out-parameters
+private class MutableStringPointer {
+    var ptr = llb_data_t()
+    init() { }
+    
+    deinit {
+        ptr.data?.deallocate()
+    }
+    
+    var msg: String? {
+        guard ptr.data != nil else { return nil }
+        return stringFromData(ptr)
+    }
+}
+
+/// Database object that defines a connection to a llbuild database
+public final class Database {
+    
+    /// Errors that can happen when opening the database or performing operations on it
+    public enum Error: Swift.Error {
+        /// If the system can't open the database, this error is thrown at init
+        case couldNotOpenDB(String)
+        /// If an operation on the database fails, this error is thrown
+        case operationDidFail(String)
+        /// If the database didn't provide an error but the operation still failed, the unknownError is thrown
+        case unknownError
+    }
+    
+    public private(set) var delegate: BuildDBDelegate
+    /// The opaque pointer to the database object
+    private var _database: OpaquePointer? = nil
+    
+    /// Initializes the build database at a given path
+    /// If the database at this path doesn't exist, it will created
+    /// If the clientSchemaVersion is different to the one in the database at this path, its content will be automatically erased!
+    public init(path: String, clientSchemaVersion: UInt32, delegate: BuildDBDelegate) throws {
+        // Safety check that we have linked against a compatibile llbuild framework version
+        if llb_get_api_version() != LLBUILD_C_API_VERSION {
+            fatalError("llbuild C API version mismatch, found \(llb_get_api_version()), expect \(LLBUILD_C_API_VERSION)")
+        }
+        
+        self.delegate = delegate
+        
+        var _delegate = llb_database_delegate_t()
+        _delegate.context = Unmanaged.passUnretained(self).toOpaque()
+        _delegate.get_key_for_id = { Database.toDB($0!).get_key_for_id($1, $2) }
+        _delegate.get_key_id = { Database.toDB($0!).get_key_id($1!) }
+        
+        let errorPtr = MutableStringPointer()
+        _database = llb_database_create(strdup(path), clientSchemaVersion, _delegate, &errorPtr.ptr)
+        
+        if let error = errorPtr.msg {
+            throw Error.couldNotOpenDB(error)
+        }
+    }
+    
+    deinit {
+        llb_database_destroy(_database)
+    }
+    
+    public final class BuildDB {
```

Comment:
Database.BuildDB is awkward, lets try and make a single type, and leverage Swift's support for throwing/failing initializers to capture other behavior.